### PR TITLE
ARROW-3311: [R] Wrap MemoryMappedFile class

### PR DESCRIPTION
**Pull request by romainfrancois**
The functions `read_record_batch` and `read_table` become S3 generic and methods handle: 

 - `arrow::io::RandomAccessFile` : that can be made with `file_open()` or `mmap_open()`
 - a file path (either a character scalar or a `fs::fs_path` object. In that case a `ReadableFile` stream is open, read from using the `arrow::io::RandomAccessFile` method and closed on exit. 
 - `arrow::BufferReader`
 - a raw vector, which uses the `arrow::BufferReader` method

Classes `Buffer`, `ReadableFile` and `MemoryMappedFile` are wrapped in R6 classes.

### CODE REVIEW

**Comment by romainfrancois** [comment]
@wesm I've updated this PR to follow up on the changes merged in #2711 

I think this can be reviewed as is, otherwise I'll also add commits about using streams in the other direction, e.g. write a record batch or a Table to an `OutputStream` but I guess that can be another PR.

**Comment by wesm** [comment]
OK, I will review this sometime today

File: r/R/io.R [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  )
+)
+
+`arrow::io::MemoryMappedFile` <- R6Class("arrow::io::MemoryMappedFile", inherit = `arrow::io::RandomAccessFile`,
+  public = list(
+    Close = function() io___MemoryMappedFile__Close(self),
+    Tell = function() io___MemoryMappedFile__Tell(self),
+    Seek = function(position) io___MemoryMappedFile__Seek(self, position),
+    Resize = function(size) io___MemoryMappedFile__Resize(self, size),
+    Read = function(nbytes) `arrow::Buffer`$new(io___Readable__Read(self, nbytes))
```
Comment (wesm) [inline r/R/io.R]: Does R6 support diamond inheritance? You probably want to follow the Arrow class hierarchy and put `Close`, `Tell`, and `Seek` in a base class, and `Read` on `InputStream`

File: r/R/io.R [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  )
+)
+
+`arrow::io::ReadableFile` <- R6Class("arrow::io::ReadableFile", inherit = `arrow::io::RandomAccessFile`,
+  public = list(
+    Close = function() io___ReadableFile__Close(self),
+    Tell = function() io___ReadableFile__Tell(self),
+    Seek = function(position) io___ReadableFile__Seek(self, position),
```
Comment (wesm) [inline r/R/io.R]: Same question here re: inheritance

File: r/src/buffer.cpp [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+// [[Rcpp::export]]
+int64_t Buffer__size(const std::shared_ptr<arrow::Buffer>& buffer) {
+  return buffer->size();
+}
+
+// [[Rcpp::export]]
+std::shared_ptr<arrow::Buffer> Buffer__parent(
+    const std::shared_ptr<arrow::Buffer>& buffer) {
+  return buffer->parent();
+}
```
Comment (wesm) [inline r/src/buffer.cpp]: Maybe better to not wrap this method unless you need it. I've been considering changing the way that slice references work. Can always remove this later, too

**Review by wesm — COMMENTED** [review]
This looks reasonable. I noted a couple typos, and left some questions about handling methods that are part of the virtual interfaces in arrow/io/interfaces.h. Let me know what (if anything) you want to do there in this patch, but we can always manage improving that code in a follow up patch

File: r/R/buffer.R [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+# specific language governing permissions and limitations
+# under the License.
+
+#' @include R6.R
+#' @include enums.R
+
+`arrow::Buffer` <- R6Class("arrow::Buffer", inherit = `arrow::Object`,
+  public = list(
+    is_mutable = function() Buffer__is_mutable(self),
+    ZeroPassing = function() Buffer__ZeroPadding(self),
```
[the remark this answers is not in the document]
Comment (romainfrancois) [reply r/R/buffer.R]: Thanks.

File: r/R/io.R (hunk 1 above)
Comment (romainfrancois) [reply r/R/io.R]: I don't think so, ping @wch.  https://github.com/r-lib/R6/issues/9

File: r/src/buffer.cpp (hunk 3 above)
Comment (romainfrancois) [reply r/src/buffer.cpp]: Was mostly a because it's there, we don't really need it.

File: r/R/io.R (hunk 1 above)
Comment (wesm) [reply r/R/io.R]: I see. Well you could implement a non-diamond inheritance hierarchy for R and at least maximize the amount of code sharing. I guess there might be the issue that the incoming shared_ptr type to each function is different, so perhaps it's not worth the effort as long as you internally dispatch to a primary implementation that uses the lowest-level virtual interface

**Comment by romainfrancois** [comment]
I'm adding a few things to it, i.e. something to replace #2741 so that we can do `read_record_batch` on a  raw vector.

**Comment by wesm** [comment]
OK, I'll return in a few hours to merge. thanks!

**Comment by romainfrancois** [comment]
@wesm I think it's good now. 

So we have both `read_table` and `read_record_batch` that are generic, with implementations for various streams: 

```r
#' Read an arrow::Table from a stream
#'
#' @param stream stream. Either a stream created by [file_open()] or [mmap_open()] or a file path.
#'
#' @export
read_table <- function(stream){
  UseMethod("read_table")
}

#' @export
read_table.character <- function(stream){
  assert_that(length(stream) == 1L)
  read_table(fs::path_abs(stream))
}

#' @export
read_table.fs_path <- function(stream) {
  stream <- file_open(stream); on.exit(stream$Close())
  read_table(stream)
}

#' @export
`read_table.arrow::io::RandomAccessFile` <- function(stream) {
  `arrow::Table`$new(read_table_RandomAccessFile(stream))
}

#' @export
`read_table.arrow::io::BufferReader` <- function(stream) {
  `arrow::Table`$new(read_table_BufferReader(stream))
}

#' @export
`read_table.raw` <- function(stream) {
  stream <- buffer_reader(stream); on.exit(stream$Close())
  read_table(stream)
}
```

so @javierluraschi reading from a raw vector is just: 

```r
bytes <- ... # from somewhere
batch <- read_record_batch(bytes)
```

This, I believe is more coherent with the rest of the api than #2741

My next move is to do the same for output streams, i.e. instead of having the `$to_file` and `to_stream`, we'd have a `stream` generic S3 with double dispatch on what we stream (record batch or table) and where (the output stream). 

I can either do this as a follow up patch, or right here.

File: r/R/io.R (hunk 1 above)
Comment (wch) [reply r/R/io.R]: That's right, R6 doesn't have multiple inheritance.

Comment (romainfrancois) [reply r/R/io.R]: Would something like java's `implements` make sense ?

File: r/R/enums.R [hunk 5]
```diff
@@ -60,3 +60,9 @@ StatusCode <- enum("arrow::StatusCode",
   PlasmaObjectExists = 20L, PlasmaObjectNonexistent = 21L,
   PlasmaStoreFull = 22L, PlasmaObjectAlreadySealed = 23L
 )
+
+#' @rdname DataType
+#' @export
+FileMode <- enum("arrow::io::FileMode",
+  READ = 0L, WRITE = 1L, READWRITE = 2L
+)
```
Comment (wesm) [inline r/R/enums.R]: Is there a way to get these enums from the C++ headers rather than hard-coding the values here?

File: r/src/RecordBatch.cpp [hunk 6]
```diff
@@ -63,17 +63,26 @@ List RecordBatch__to_dataframe(const std::shared_ptr<arrow::RecordBatch>& batch)
 }
 
 // [[Rcpp::export]]
-std::shared_ptr<arrow::RecordBatch> read_record_batch_(std::string path) {
-  std::shared_ptr<arrow::io::ReadableFile> stream;
+std::shared_ptr<arrow::RecordBatch> read_record_batch_RandomAccessFile(
```
Comment (wesm) [inline r/src/RecordBatch.cpp]: We need to revisit these APIs -- read_record_batch_RandomAccessFile and read_record_batch_BufferReader do different things. https://issues.apache.org/jira/browse/ARROW-3498

File: r/src/RecordBatch.cpp [hunk 7]
```diff
@@ -63,17 +63,26 @@ List RecordBatch__to_dataframe(const std::shared_ptr<arrow::RecordBatch>& batch)
...
 
   std::shared_ptr<arrow::RecordBatch> batch;
   R_ERROR_NOT_OK(rbf_reader->ReadRecordBatch(0, &batch));
 
-  R_ERROR_NOT_OK(stream->Close());
+  return batch;
+}
+
+// [[Rcpp::export]]
+std::shared_ptr<arrow::RecordBatch> read_record_batch_BufferReader(
```
Comment (wesm) [inline r/src/RecordBatch.cpp]: This function name does not suggest that it reads a stream (including the schema). The Python version of this does not read a stream

File: r/src/RecordBatch.cpp [hunk 8]
```diff
@@ -95,20 +104,20 @@ int RecordBatch__to_file(const std::shared_ptr<arrow::RecordBatch>& batch,
   return offset;
 }
 
-// [[Rcpp::export]]
-RawVector RecordBatch__to_stream(const std::shared_ptr<arrow::RecordBatch>& batch) {
+int64_t RecordBatch_size(const std::shared_ptr<arrow::RecordBatch>& batch) {
```
Comment (wesm) [inline r/src/RecordBatch.cpp]: Hm, "stream_size"?

File: r/src/Table.cpp [hunk 9]
```diff
@@ -78,7 +98,27 @@ std::shared_ptr<arrow::Table> read_table_(std::string path) {
 
   std::shared_ptr<arrow::Table> table;
   R_ERROR_NOT_OK(arrow::Table::FromRecordBatches(std::move(batches), &table));
-  R_ERROR_NOT_OK(stream->Close());
+
+  return table;
+}
+
+// [[Rcpp::export]]
+std::shared_ptr<arrow::Table> read_table_BufferReader(
```
Comment (wesm) [inline r/src/Table.cpp]: Same here

**Review by wesm — COMMENTED** [review]
Some minor comments about API refinement. We can address in follow up patches

**Comment by wesm** [comment]
+1

File: r/R/enums.R (hunk 5 above)
Comment (romainfrancois) [reply r/R/enums.R]: I’m not aware of a way to do that. Perhaps we can parse the headers somehow, IIRC once it’s compiled, it’s too late. @jjallaire ?

File: r/src/RecordBatch.cpp (hunk 7 above)
Comment (romainfrancois) [reply r/src/RecordBatch.cpp]: It reads a record batch from a BufferReader. 

But yeah this is likely to change once more C++ classes can be held on the r side. 

Should it just read the record batch and not the schema ?

Comment (wesm) [reply r/src/RecordBatch.cpp]: We need to go carefully through all the modes of operation when it comes to IPC, as listed in https://issues.apache.org/jira/browse/ARROW-3498. We have support for all modes in Python at the moment:

* Reading a stream
* Reading a file
* Reading a schema alone
* Reading a single record batch alone given a known schema

File: r/src/RecordBatch.cpp (hunk 8 above)
Comment (romainfrancois) [reply r/src/RecordBatch.cpp]: That makes sense. I’ve switched to the other side of the problem. I’ll keep this in mind and propose some changes.

File: r/R/io.R (hunk 1 above)
Comment (wch) [reply r/R/io.R]: @romainfrancois want to file an issue on R6? It's worth having a discussion about it.

File: r/R/enums.R (hunk 5 above)
Comment (jjallaire) [reply r/R/enums.R]: Yes, we'd have to parse the headers during some sort of R code generation phase. We don't currently have a convenient place to do this though.

Comment (wesm) [reply r/R/enums.R]: I see. It would be interesting if Rcpp could generate constant values

File: r/DESCRIPTION [hunk 10]
```diff
@@ -3,6 +3,7 @@ Title: R Integration to 'Apache' 'Arrow'
 Version: 0.0.0.9000
 Authors@R: c(
     person("Romain", "François", email = "romain@rstudio.com", role = c("aut", "cre")),
+    person("Javier", "Luraschi", email = "javier@rstudio.com", role = c("ctb")),
```
Comment (javierluraschi) [inline r/DESCRIPTION]: @romainfrancois Lol, I don't think I deserve this yet... so feel free to remove be before going to CRAN... unless I ended up contributing something more significant, in any case, thanks!

**Pull request closed without merging**